File: moderate/flavius_josephus.rb
Node = Struct.new(:value, :next)
def list_push(value, node)
  return Node.new(value, node)
end

File.open(ARGV[0]).each_line do |line|
  s = line.split(',').map(&:to_i)
  n, m = s[0], s[1]

  tail = Node.new(n - 1, nil)
  list = tail
  (n - 2).step(0, -1) { |x| list = list_push(x, list) }
  tail.next = list
  list = tail

  s = []
  n.times {
    (m-1).times { list = list.next }
    s << list.next.value
    list.next = list.next.next
  }
  puts s.join(" ")
end

File: moderate/flavius_josephus.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type node struct {
	value int
	next  *node
}

func main() {
	var n, m int
	data, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer data.Close()
	scanner := bufio.NewScanner(data)
	for scanner.Scan() {
		var t []string
		fmt.Sscanf(scanner.Text(), "%d,%d", &n, &m)

		tail := node{value: n - 1}
		list := &tail
		for i := n - 2; i >= 0; i-- {
			list = &node{i, list}
		}
		tail.next, list = list, &tail

		for i := 0; i < n; i++ {
			for j := 0; j < m-1; j++ {
				list = list.next
			}
			t = append(t, strconv.Itoa(list.next.value))
			list.next = list.next.next
		}
		fmt.Println(strings.Join(t, " "))
	}
}
